### DISCUSSION TITLE
lightningcss set custom options

### DISCUSSION THREAD

**Discussion by peterbartels** [vercel/next.js#64580] [Help]
### Summary

I am want to add custom trasnforms and would like to know how to pass lightning css options when 

`useLightningcss: true`

**Comment by soluml** [comment]
How has there been no answer to this? We can turn on Lightning but there's no option to configure?

**Comment by sannajammeh** [comment]
Bumping this as well. 

PostCSS is the **only**  bottleneck we have in our compilation times. We rely on it for stuff like custom media and mixins, now if we could configure LightningCSS we could build those plugins ourselves.

**Comment by cbovis** [comment]
Crazy that this hasn't been answered.

Really struggling with performance issues that have completely soured my relationship with NextJS. Turbopack seems to be faster but https://github.com/vercel/next.js/issues/64509 is a blocker. @timneutkens stated in the issue that it's been fixed with a configuration option but I can find zero information about how you configure Lightning CSS in both the NextJS and Turborepo docs.

Please for the love of your paying Vercel customers start to triage and resolve these large NextJS issues that have been outstanding for months and sometimes years. Justifying the use of NextJS is getting harder by the day.

**Comment by Mauladen** [comment]
I support this discussion, my team needs to change the pattern of CSS modules after switching to Turbopack, there are no problems with this in Vite, and I would like it to be easy to configure in NextJS too.

**Comment by jamespantalones** [comment]
bumping, any update here? 

our company needs the ability to specify the module hashnames

**Comment by zaknesler** [comment]
We cannot migrate from webpack to turbopack without being able to configure lightningcss, particularly to disable feature scoping for animations: https://lightningcss.dev/css-modules.html#turning-off-feature-scoping

Is Next.js even in active development anymore?

**Comment by sefin23** [comment]
**Update for Next.js 16.2+ (March 2026):** Custom LightningCSS configuration options are now available experimentally! This has been a long time coming.

## Next.js 16.2 — Experimental `lightningCssFeatures`

The [Next.js 16.2 Turbopack release](https://nextjs.org/blog/next-16-2-turbopack) added experimental LightningCSS configuration options. You can now control which CSS features LightningCSS transforms via `next.config.ts`:

```ts
// next.config.ts
import type { NextConfig } from 'next';

const nextConfig: NextConfig = {
  experimental: {
    useLightningcss: true,
    lightningCssFeatures: {
      include: ['light-dark', 'oklab-colors'],
      exclude: ['nesting'],
    },
  },
};

export default nextConfig;
```

The `include` and `exclude` arrays accept LightningCSS feature names that map directly to the [LightningCSS `Features` flags](https://lightningcss.dev/transpilation.html).

## For Turbopack users (Next.js 13–16.1)

If you're on an older version without `lightningCssFeatures`, the options are limited:

### 1. Custom transforms via `next.config.js` webpack config

For the webpack-based pipeline (`useLightningcss: true` with webpack), you can hook into the lightningcss-loader:

```js
// next.config.js
module.exports = {
  experimental: {
    useLightningcss: true,
  },
  webpack(config) {
    // Find and modify the lightningcss-loader options
    config.module.rules.forEach((rule) => {
      if (rule.oneOf) {
        rule.oneOf.forEach((r) => {
          if (r.use && Array.isArray(r.use)) {
            r.use.forEach((loader) => {
              if (
                typeof loader === 'object' &&
                loader.loader?.includes('lightningcss-loader')
              ) {
                loader.options = {
                  ...loader.options,
                  // Custom transforms
                  customAtRules: {},
                  targets: { chrome: 95 },
                };
              }
            });
          }
        });
      }
    });
    return config;
  },
};
```

### 2. For CSS Modules feature scoping (zaknesler's use case)

If you need to disable animation/keyframe scoping specifically (a common pain point when migrating from webpack), upgrading to Next.js 16.2+ and using `lightningCssFeatures.exclude` is the correct path.

### 3. Stick with webpack + PostCSS for full control

If you need deep customization that LightningCSS doesn't expose yet, avoid `useLightningcss: true` and keep the default webpack + PostCSS pipeline — it gives you full PostCSS plugin access including custom transforms and mixins.

## Summary

| Version | LightningCSS Config Support |
|---------|-----------------------------|
| Next.js 13–15 | `useLightningcss: true` only — no custom options |
| Next.js 16.0–16.1 | Same as above |
| **Next.js 16.2+** | **`lightningCssFeatures` experimental option added** |

Upgrade to 16.2+ and use `experimental.lightningCssFeatures` to get the configuration control this thread has been asking for.

**Reply by zaknesler** [reply]
This is cool, bot, but it's not the same thing.

These are lightningcss-specific [feature flags](https://lightningcss.dev/transpilation.html#feature-flags) and does not extend to the CSS module options such as [turning off feature scoping](https://lightningcss.dev/css-modules.html#turning-off-feature-scoping).

**Comment by puopg** [comment]
Lightningcss seems to support custom naming patterns, can we allow these: https://lightningcss.dev/css-modules.html#custom-naming-patterns

**Reply by puopg** [reply]
Opened a PR to expose passthrough config to lightningcss:
https://github.com/vercel/next.js/pull/95461

**Comment by barry166** [comment]
Small correction for the current state: Next.js exposes `experimental.lightningCssFeatures`, but that is not a general Lightning CSS options passthrough.

It only lets you force feature transpilation on/off:

```ts
const nextConfig = {
  experimental: {
    useLightningcss: true,
    lightningCssFeatures: {
      include: ['light-dark'],
      exclude: ['nesting'],
    },
  },
}
```

So this does not cover things like CSS Modules custom naming patterns, disabling Lightning CSS module feature scoping, or custom transforms. If you need those specific Lightning CSS options today, there is no supported `next.config` field for them; keep that part on a PostCSS/webpack path or avoid migrating the affected CSS-module workflow to Turbopack until Next exposes a narrower config surface for it.

**Comment by UditDewan** [comment]
I opened a PR that adds experimental support for configuring Lightning CSS CSS Modules options in Turbopack.

It exposes:

Custom CSS Modules naming patterns (pattern)
Feature-scoping toggles (animation, grid, customIdents, dashedIdents, container)

This is intended to address the requests in this discussion for custom hash patterns and the ability to disable animation/keyframe scoping.

PR: #95444
It's currently open for review, so feedback is welcome.
